File: app/src/main/java/org/tasks/tags/TagPickerViewHolder.java
package org.tasks.tags;

import android.content.Context;
import android.view.View;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnCheckedChanged;
import butterknife.OnClick;
import kotlin.jvm.functions.Function2;
import org.tasks.R;
import org.tasks.data.TagData;
import org.tasks.tags.CheckBoxTriStates.State;
import org.tasks.themes.DrawableUtil;

public class TagPickerViewHolder extends RecyclerView.ViewHolder {

  private final Context context;
  private final Function2<TagData, Boolean, State> callback;

  @BindView(R.id.text)
  TextView text;

  @BindView(R.id.checkbox)
  CheckBoxTriStates checkBox;

  private TagData tagData;

  TagPickerViewHolder(
      Context context, @NonNull View view, Function2<TagData, Boolean, State> callback) {
    super(view);
    this.callback = callback;

    ButterKnife.bind(this, view);

    this.context = context;
  }

  @OnClick(R.id.tag_row)
  void onClickRow() {
    if (tagData.getId() == null) {
      callback.invoke(tagData, true);
    } else {
      checkBox.toggle();
    }
  }

  @OnCheckedChanged(R.id.checkbox)
  void onCheckedChanged() {
    State newState = callback.invoke(tagData, checkBox.isChecked());
    updateCheckbox(newState);
  }

  public void bind(
      @NonNull TagData tagData, int color, @Nullable Integer icon, State state) {
    this.tagData = tagData;
    if (tagData.getId() == null) {
      text.setText(context.getString(R.string.create_new_tag, tagData.getName()));
      icon = R.drawable.ic_outline_add_24px;
      checkBox.setVisibility(View.GONE);
    } else {
      text.setText(tagData.getName());
      if (state == State.CHECKED) {
        checkBox.setChecked(true);
      } else {
        updateCheckbox(state);
      }
      if (icon == null) {
        icon = R.drawable.ic_outline_label_24px;
      }
    }
    DrawableUtil.setLeftDrawable(context, text, icon);
    DrawableUtil.setTint(DrawableUtil.getLeftDrawable(text), color);
  }

  private void updateCheckbox(State state) {
    checkBox.setState(state, false);
    checkBox.setVisibility(View.VISIBLE);
  }
}

File: app/src/main/java/org/tasks/tags/TagPickerViewHolder.kt
package org.tasks.tags

import android.content.Context
import android.view.View
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import org.tasks.R
import org.tasks.data.TagData
import org.tasks.databinding.RowTagPickerBinding
import org.tasks.themes.DrawableUtil

class TagPickerViewHolder internal constructor(
    private val context: Context,
    binding: RowTagPickerBinding,
    private val callback: (TagData, TagPickerViewHolder) -> Unit
) : RecyclerView.ViewHolder(binding.root) {

    val isChecked: Boolean
        get() = checkBox.isChecked

    private val text: TextView = binding.text
    private val checkBox: CheckBoxTriStates = binding.checkbox.apply {
        setOnCheckedChangeListener { _, _ ->  onCheckedChanged() }
    }
    private var tagData: TagData? = null

    private fun onClickRow() {
        if (tagData!!.id == null) {
            callback(tagData!!, this)
        } else {
            checkBox.toggle()
        }
    }

    private fun onCheckedChanged() {
        callback(tagData!!, this)
    }

    fun bind(
            tagData: TagData, color: Int, icon: Int?, state: CheckBoxTriStates.State) {
        var icon = icon
        this.tagData = tagData
        if (tagData.id == null) {
            text.text = context.getString(R.string.create_new_tag, tagData.name)
            icon = R.drawable.ic_outline_add_24px
            checkBox.visibility = View.GONE
        } else {
            text.text = tagData.name
            if (state == CheckBoxTriStates.State.CHECKED) {
                checkBox.isChecked = true
            } else {
                updateCheckbox(state)
            }
            if (icon == null) {
                icon = R.drawable.ic_outline_label_24px
            }
        }
        DrawableUtil.setLeftDrawable(context, text, icon)
        DrawableUtil.setTint(DrawableUtil.getLeftDrawable(text), color)
    }

    fun updateCheckbox(state: CheckBoxTriStates.State) {
        checkBox.setState(state, false)
        checkBox.visibility = View.VISIBLE
    }

    init {
        binding.tagRow.setOnClickListener { onClickRow() }
    }
}